File: src/julia/ising_SW.jl
require("observable.jl")
require("lattice.jl")

using UnionFinds

type cluster
  spin :: Int
  size :: Int
end

function ising_SW(lat::Lattice, T::Float64, Sweeps::Int, Thermalization::Int)
  nsites = num_sites(lat)
  nbonds = num_bonds(lat)
  spins = ones(Int,nsites)

  p = 1.0-exp(-2.0/T)

  function update()
    uf = UnionFind(nsites)
    for bond in 1:nbonds
      s1,s2 = source(lat, bond), target(lat, bond)
      if spins[s1] == spins[s2] && rand() < p
        unify(uf, s1,s2)
      end
    end
    clusters = Dict{Int,cluster}()
    for site in 1:nsites
      r = root(uf,site)
      if !has(clusters,r)
        clusters[r] = cluster(2(rand(0:1))-1, 0)
      end
      cl = clusters[r]
      spins[site] = cl.spin
      cl.size += 1
    end
    return (Float64)[ square(cl[2].size/nsites) for cl in clusters]
  end

  obs = ObservableSet()
  add(obs,"Magnetization")
  add(obs,"Magnetization^2")
  add(obs,"Magnetization^4")
  add(obs,"Energy")
  add(obs,"Energy^2")

  function measure(ar::Vector)
    m2 = 0.0
    m4 = 0.0

    for i in 1:length(ar)
      mi = ar[i]
      m2 += mi
      m4 += square(mi)
      for j in 1:i-1
        m4 += 6.0*mi*ar[j]
      end
    end

    add(obs["Magnetization"], 0.0)
    add(obs["Magnetization^2"], m2)
    add(obs["Magnetization^4"], m4)
    #add(obs["Energy"], en)
    #add(obs["Energy^2"], square(en))
  end

  for mcs in 1:Thermalization
    update()
  end
  for mcs in 1:Sweeps
    measure(update())
  end

  return obs
end

L = 16
lat = square_lattice(L)
Sweeps = 8192
Thermalization = Sweeps>>3

for T in 2.0:0.1:2.4
  observables = ising_SW(lat, T, Sweeps, Thermalization)
  for obs in observables
    open("$(obs[1])_$L.dat", "a") do os
      write(os, "$T $(mean(obs[2])) $(error(obs[2]))\n")
    end
  end
end

File: python/ising_SW.py
import math
import numpy as np
import numpy.random as random

execfile('observable.py')
execfile('unionfind.py')
execfile('square.py')

class SW:
  def __init__(self, L, T):
    ## parameters
    self.L = L
    self.lattice = Square(L)
    self.T = T
    self.beta = 1.0/T

    ## configurations
    self.spins = np.ones(self.num_site(), dtype=np.int)
    self.mag = self.num_site()
    self.ene = -self.num_bond()

    ## algorithm parameter
    self.prob = 1.0-math.exp(-2*self.beta)
    self.UF = UnionFind(self.num_site())

  ## delegate to lattice
  def num_site(self) : return self.lattice.num_site()
  def num_bond(self) : return self.lattice.num_bond()
  def neighbors(self, site) : return self.lattice.neighbors(site)
  def site2coord(self, site) : return self.lattice.site2coord(site)
  def coord2site(self, x,y) : return self.lattice.coord2site(x,y)

  ## delegate to UnionFind
  def unify(self, lsite, rsite) : return self.UF.unify(lsite, rsite)
  def cluster_id(self, site) : return self.UF.cluster_id(site)

  def update(self):
    self.UF.reset(self.num_site())

    num_activated = 0

    for site in xrange(self.num_site()):
      x,y = self.site2coord(site)
      spin = self.spins[site]
      for neighbor in self.neighbors(site)[0:2]:
        if self.spins[neighbor] == spin and random.rand() < self.prob :
          self.unify(site, neighbor)
          num_activated += 1
    num_clusters = self.UF.clusterize()
    cl_spins = random.randint(2, size=num_clusters)*2-1
    cl_sizes = np.zeros(num_clusters)

    for site in xrange(self.num_site()):
      cid = self.cluster_id(site)
      self.spins[site] = cl_spins[cid]
      cl_sizes[cid] += 1.0

    am = np.sum(self.spins)
    am *= 1.0/self.num_site()
    am = abs(am)
    cl_sizes *= 1.0/self.num_site()
    cl_sizes *= cl_sizes
    m2 = np.sum(cl_sizes)
    cs = np.cumsum(cl_sizes)
    m4 = 12*np.dot(cl_sizes[1:], cs[:-1]) + np.dot(cl_sizes,cl_sizes)

    ene = self.num_bond() - (1.0+1.0/math.tanh(self.beta))*num_activated
    ene /= self.num_bond()

    return am, m2, m4, ene

  def run(self, MCS=8192, Thermalization=1024):
    for mcs in xrange(Thermalization):
      self.update()
    absmag = Observable()
    mag2 = Observable()
    mag4 = Observable()
    sus = Observable()
    ene = Observable()
    for mcs in xrange(MCS):
      am,m2, m4, en = self.update()
      chi = m2 * self.beta/self.num_site()
      absmag.add(am)
      mag2.add(m2)
      mag4.add(m4)
      sus.add(chi)
      ene.add(en)
    return absmag, mag2, mag4, sus, ene
